datatypes-jws/src/commonMain/kotlin/at/asitplus/crypto/datatypes/jws/JwsSigned.kt

```diff
@@ -29,31 +33,38 @@ data class JwsSigned(
 
         if (header != other.header) return false
         if (!payload.contentEquals(other.payload)) return false
-        if (!signature.contentEquals(other.signature)) return false
-        return plainSignatureInput == other.plainSignatureInput
+        return signature == other.signature
     }
 
     override fun hashCode(): Int {
         var result = header.hashCode()
         result = 31 * result + payload.contentHashCode()
-        result = 31 * result + signature.contentHashCode()
-        result = 31 * result + plainSignatureInput.hashCode()
+        result = 31 * result + signature.hashCode()
         return result
     }
 
     companion object {
         fun parse(it: String): JwsSigned? {
             val stringList = it.replace("[^A-Za-z0-9-_.]".toRegex(), "").split(".")
             if (stringList.size != 3) return null.also { Napier.w("Could not parse JWS: $it") }
-            val headerInput = stringList[0].decodeToByteArrayOrNull(Base64Strict)
+            val headerInput = stringList[0].decodeToByteArrayOrNull(Base64UrlStrict)
                 ?: return null.also { Napier.w("Could not parse JWS: $it") }
             val header = JwsHeader.deserialize(headerInput.decodeToString())
                 ?: return null.also { Napier.w("Could not parse JWS: $it") }
-            val payload = stringList[1].decodeToByteArrayOrNull(Base64Strict)
+            val payload = stringList[1].decodeToByteArrayOrNull(Base64UrlStrict)
                 ?: return null.also { Napier.w("Could not parse JWS: $it") }
-            val signature = stringList[2].decodeToByteArrayOrNull(Base64Strict)
-                ?: return null.also { Napier.w("Could not parse JWS: $it") }
-            return JwsSigned(header, payload, signature, "${stringList[0]}.${stringList[1]}")
+            val signature = stringList[2].decodeToByteArrayOrNull(Base64UrlStrict)
+                ?.let { it1 ->
+                    when (header.algorithm) {
+                        JwsAlgorithm.ES256, JwsAlgorithm.ES384, JwsAlgorithm.ES512 -> CryptoSignature.EC(it1)
+                        else -> CryptoSignature.RSAorHMAC(it1)
+                    }
+                } ?: return null.also { Napier.w("Could not parse JWS: $it") }
+
+            return JwsSigned(header, payload, signature)
         }
     }
-}
\ No newline at end of file
+}
+
+fun prepareJwsSignatureInput(header: JwsHeader, payload: ByteArray): String =
```

Comment:
If `header` now gets serialized differently than it was parsed, the signature input is different, and the signature will not match ...

Reply:
You need to help me understand…

Reply:
The JWS compact serialization (the string) gets deserialized into an object of `JwsSigned`, and then its signature gets verified (e.g. in https://github.com/a-sit-plus/kmm-vc-library/blob/main/vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/jws/JwsService.kt#L219). This signature verification should use the string values as they've been serialized (i.e. in the string input), so that the signature will match. But now this original input string is missing from `JwsSigned`, so I'll guess, the signature can't be verified everytime (e.g. when `header.serialize()` now yields a slightly different representation)